### PR TITLE
Fix Color.lerp to interpolate linearly

### PR DESCRIPTION
**Pull request by mattsarett**
https://github.com/flutter/flutter/issues/10655

### CODE REVIEW

**Comment by mattsarett** [comment]
@brianosman

**Comment by Hixie** [comment]
This seems remarkably more expensive than what we had before. Is the improvement that noticeable?

cc @abarth

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -8,6 +8,29 @@ Color _scaleAlpha(Color a, double factor) {
   return a.withAlpha((a.alpha * factor).round());
 }
 
+/// Apply to canonical sRGB transfer function to convert to linear values.
+double _srgbToLinear(int color) {
+  assert(0 <= color && color <= 255);
+
+  double x = color * (1.0 / 255.0);
+  if (x <= 0.04045) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: nit: no braces

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -8,6 +8,29 @@ Color _scaleAlpha(Color a, double factor) {
...
+  }
+
+  return math.pow((x + 0.055) * (1.0 / 1.055), 2.4);
+}
+
+/// Apply the canonical inverse sRGB transfer function to convert from linear values.
+int _srgbFromLinear(double x) {
+  assert(0.0 <= x && x <= 1.0);
+
+  if (x <= 0.0031308) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: nit: no braces

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -8,6 +8,29 @@ Color _scaleAlpha(Color a, double factor) {
...
+    return (255.0 * (x * 12.92)).toInt();
+  }
+
+  return (255.0 * (1.055 * math.pow(x, (1.0 / 2.4)) - 0.055)).toInt();
+}
+
```
Comment (Hixie) [inline lib/ui/painting.dart]: nit: I'd probably just put these functions in the Color class itself as static methods just before or after the lerp method itself, to make it easier to see them when reading the code.

**Comment by Hixie** [comment]
This will need a test.

File: lib/ui/painting.dart (hunk 1 above)
Comment (mattsarett) [inline lib/ui/painting.dart]: Done

File: lib/ui/painting.dart (hunk 2 above)
Comment (mattsarett) [inline lib/ui/painting.dart]: Done

File: lib/ui/painting.dart (hunk 3 above)
Comment (mattsarett) [inline lib/ui/painting.dart]: Done

**Comment by mattsarett** [comment]
The test is here:
https://github.com/flutter/flutter/pull/10699

I would say that the difference is very noticeable, particularly when using the center of the gradient as a measure of how it "should" look.

But haha yes it is very expensive.  In general, I would say that this type of operation might be something better deferred to Skia rather than exposed in Dart in this way.  When possible, we take advantage of GPU hardware or optimized CPU code to apply the transfer function.

It looks like Flutter's Color class is a 32-bit int that corresponds directly to an SkColor.  Since SkColor is defined to be sRGB, operations on Colors will need to take this into account.

**Comment by Hixie** [comment]
That's a manual test, but algorithmic changes like this should also have automated tests so that we don't regress it (we're not going to be looking at the manual tests with every checkin).

We can't really defer to skia for this because that would involve a language jump, which is also expensive and would almost certainly swamp out any benefit to doing things in skia.

We run this code many, many times per frame when there's a theme animation. This is performance-critical code.

**Comment by mattsarett** [comment]
Sounds good, I'll write an automated test.

There are various things we can consider on the performance side.  One thing that I think would be really cool is if the Color class stored r, g, b, a as 4 linear floats (with a color gamut).  If we are no longer using a packed 32-bit integer, there is no reason to use a transfer function encoding to preserve precision.  Linear floats provide plenty of precision.

Clients would get colors that are:
(1) High precision
(2) Linear (easier to reason to about)
(3) Wide gamut (right now we can only represent colors in the sRGB color gamut)

It would be important to clearly define that the Color components are linear and how the color gamut is specified (probably sRGB by default).

**Comment by Hixie** [comment]
Changing colors from a 32bit integer to 4 floats is unlikely to have a particularly positive impact on performance either...

FWIW I did a quick benchmark of this patch and it regresses Color.lerp performance about ten-fold.

What we could do is have Color.lerp do the cheap behaviour like today, and add a Color.linearLerp that uses the code you have here.

Another option might be to have an approximation of this algorithm that is fast. Currently we pay the cost of a few branches plus three floating point arithmetic operations for each color channel. If we could get something in that ballpark that approximates linear interpolation better, that would be great.

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -154,6 +154,27 @@ class Color {
...
+    double x = color * (1.0 / 255.0);
+    if (x <= 0.04045)
+      return x * (1.0 / 12.92);
+
+    return math.pow((x + 0.055) * (1.0 / 1.055), 2.4);
+  }
+
+  /// Apply the canonical inverse sRGB transfer function to convert from linear values.
+  static int _srgbFromLinear(double x) {
+    assert(0.0 <= x && x <= 1.0);
```
Comment (Hixie) [inline lib/ui/painting.dart]: BTW this assumption is not valid. In particular, the `t` passed to `lerp` could be outside the range 0.0-1.0. (I discovered this when running this code in a quick benchmark.)

**Comment by brianosman** [comment]
The cheapest approximation is going to be x^2 and sqrt. That's similarly shaped, and may be good enough for this use-case (especially if we also provide a more exact fallback). The other option is to explore fitting a function to the lerp behavior - we've done lots of curve fitting for the srgb <-> linear functions, but in this case it's a slightly different beast. There might be a much faster approximation that bypasses the need to go to linear at all.

**Comment by Hixie** [comment]
I'd have to benchmark it, but I doubt `sqrt` will be sufficiently faster than `pow` to let us use this on the lerp that's used for animations.

**Comment by Hixie** [comment]
I think the path forward here is to provide this lerp as an option called `Color.linearLerp` or something, and not use it in features such as theme animations where it would be too expensive.

I'm closing this PR for now since it hasn't been touched in 10 days , but please don't hesitate to reopen it or open a new one if there is an updated PR to review!

**Comment by mattsarett** [comment]
Closing this for now sgtm.

@brianosman will possibly follow up on this.  I think there is a good case to be made for storing color as 4 linear floats.
***Float gives us plenty of precision, meaning we can drop the transfer function encoding and store values linearly.  This makes interpolation trivial.  And even faster because we don't need to do the float<->int.
***Floats allows us to go above 1.0f or below 0.0f, so we can represent colors outside of the sRGB gamut (most mobile devices have displays that are capable to rendering these colors, so it is nice for developers to be able to specify them).
***The downside is that Color gets 4x larger.

**Pull request closed without merging**